### ISSUE TITLE
Updating app version breaks authentication

### ISSUE THREAD

**Issue by brotskydotcom2** [clickonetwo/whisper-server#2]
Many users report that, when they update their app version, they can no longer authenticate with the server.  On iOS deleting and reinstalling the app fixes the problem, probably because the app gets a new client ID and then the server accepts its locally-generated secret.  But this is harder to fix on macOS, because you have to know to delete `~/Library/Containers/Whisper` rather than just the app itself.

What's supposed to happen is that the app should notify its last secret with every APNS registration, and that secret should be accepted by the server.  But apparently that's not always working when the app version changes.

**Comment by brotskydotcom2** [comment]
So log investigation suggests that this is a mac-specific problem related to the multiple deliveries of APNS registrations to the client, and the client then posting multiple notifications to the server.  The second notification with the *old* client secret arrives just after the confirmation of the receipt of the new secret by the client, at which point the server thinks the old secret has changed and issues a new secret *again*.  But then on the client the old secret has changed to be the correct prior secret, so when the client goes to authenticate its "last secret" doesn't match the server's "last secret" (which was replaced with the old client last secret).

Basically, there's a race condition going on.  And deleting the client preferences might not even fix that.  Probably I need to do some sort of race monitoring where I anticipate that the last secret sent by the client might actually predate the client's send of the confirmation that it got the new secret.  I don't like the idea of making the client swallow the repeat notifications, because that kind of logic is often buggy and changing the client is hard.

**Comment by brotskydotcom2** [comment]
So the first fix didn't quite take!  We have to save the posted secret whenever we are going to update the client's secret, and we weren't handling all the cases!

**Comment by brotskydotcom2** [comment]
Still broken in the v2 server.  The root of the problem is that Mac clients get two of these notifications within a millisecond or two of each other, resulting in two identical calls that may arrive at the server and be processed simultaneously.  This has to be detected synchronously on the server side (not relying on an async database call) because there's no way otherwise to be sure which call arrives first and which should be discarded.
